lib/Sema/TypeCheckDecl.cpp

```diff
@@ -5603,6 +5606,177 @@ static Optional<ObjCReason> shouldMarkClassAsObjC(TypeChecker &TC,
+static bool checkNominalDeclCircularity(Decl *decl,
+  llvm::SmallPtrSetImpl<NominalTypeDecl *> &known,
+  Type typeBase = nullptr) {
+  if (auto s = dyn_cast<StructDecl>(decl)) {
+    if (checkStructDeclCircularity(s, known, typeBase)) {
+      return true;
+    }
+  } else if (auto e = dyn_cast<EnumDecl>(decl)) {
+    if (checkEnumDeclCircularity(e, known, typeBase)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+
+// break down type that "contains" other types and check them each.
+static bool deconstructTypeForDeclCircularity(Type type,
+  llvm::SmallPtrSetImpl<NominalTypeDecl *> &known) {
+
+  if (auto optional = dyn_cast<OptionalType>(type.getPointer())) {
+    // unwarp optionals, check again.
+    if (deconstructTypeForDeclCircularity(optional->getBaseType(), known)) {
+      return true;
+    }
+  } else if (auto tuple = type->getAs<TupleType>()) {
+    // check each element in tuple
+    for (auto Elt: tuple->getElements()) {
+      if (deconstructTypeForDeclCircularity(Elt.getType(), known)) {
+        return true;
+      }
+    }
+  } else {
+    if (auto decl = type->getAnyNominal()) {
+      // An ciruclarity is found! Stop the checking from here on out.
+      if (known.count(decl)) {
+        return true;
+      }
+
+      if (auto bgt = type->getAs<BoundGenericType>()) {
+        // for bound generic types, apply all generic parameter substitutions
+        // and check if its declarition contains any circularity given what
+        // has been discovered so far.
+        if (checkNominalDeclCircularity(decl, known, type)) {
+          return true;
+        }
+        // if a generic argument is itself a bound generic type, recurse.
+        for (auto arg: bgt->getGenericArgs()) {
+          if (arg->getAs<BoundGenericType>()) {
+            if (deconstructTypeForDeclCircularity(arg, known)) {
+              return true;
+            }
+          }
+        }
+      } else {
+        // check nominal type for cirularities given what has been discovered
+        // so far.
+        if (checkNominalDeclCircularity(decl, known)) {
+          return true;
+        }
+      }
+    }
+
+  }
+
+  return false;
+}
+
+
+static bool checkStructDeclCircularity(StructDecl *S,
+  llvm::SmallPtrSetImpl<NominalTypeDecl *> &known,
+  Type baseType) {
+
+  auto st = baseType ?: S->getDeclaredTypeInContext();
+
+  // Having a base type means this function is checking a generic type
+  // parameter. This declaration should not be a starting point for
+  // circularity. (e.g. in `let s: [[Int]]` the inner array should be
+  // considered legal.
+  if (!baseType) {
+    known.insert(S);
+  }
+  for (auto field: S->getStoredProperties()) {
+    if (auto vd = dyn_cast<VarDecl>(field)) {
+      // skip uninteresting fields.
+      if (vd->isStatic() || !vd->hasStorage() || !vd->hasType()) {
+        continue;
+      }
+
+      auto vdt = st->getTypeOfMember(S->getModuleContext(), vd, nullptr);
```

Comment:
You should be doing 'baseType->getTypeOfMember()' here -- the idea is to apply the baseType substitutions to the member's type